### ISSUE TITLE
addressfield always “required”

### ISSUE THREAD

**Issue by ghost** [backdrop-contrib/addressfield#6]
an addressfield always behave as if it were marked "required", whatever the state of the "required" checkbox; A workaround is to check "Make all fields optional", but this has the side effect that the field is never empty, and if a geocoder is around, it locates your object in the center of the default country;-(

**Comment by ghost** [comment]
Found the cause of the issue : Using a "Default country" makes the field "non empty", even if the user did not type anything...
Specifying Default country as "- None -" solves the problem, but is not very user friendly - and narrowing the choice of "Available countries" is a nightmare (it's a lot easier to manually edit `field.instance.node.article.field_address.json`)...

**Comment by jenlampton** [comment]
I agree this is a strange behavior. There are multiple similar issues in the Drupal queue about addresses behaving counter to how people expect:
- https://www.drupal.org/node/1844846
- https://www.drupal.org/node/1852510
- https://www.drupal.org/node/2479995
- https://www.drupal.org/node/968112
- https://www.drupal.org/node/1443674
- https://www.drupal.org/node/2282817

We should perhaps review those and see if we can come up with a solution.

**Comment by alanmels** [comment]
The fix doesn't have to be complex by leaving everything intact as is, but just checking if the whole addressfield was marked as required by user or not and depending on their choice make the addressfield elements required or not.

The addressfield that we use, unfortunately, is getting further and further away from this one as another suggestion of mine on https://github.com/backdrop-contrib/addressfield/issues/35 was never accepted, so I hope this simple fix will go in soon.

**Comment by alanmels** [comment]
Will this ever be considered or should we just keep adding changes to https://github.com/alanmels/addressfield/tree/1.x-1.x that is diverging further and further?

**Comment by argiepiano** [comment]
@alanmels I'm not sure if I understand the problem. There is a checkbox in the UI that allows you to make all fields optional:

<img width="573" alt="Screen Shot 2025-01-09 at 8 55 27 PM" src="https://github.com/user-attachments/assets/6478f06b-7c30-4da8-ae3d-a3abcfd3560d" />

Before checking that box (shows required subfields despite the fact that the field is not required):
<img width="900" alt="Screen Shot 2025-01-09 at 8 56 40 PM" src="https://github.com/user-attachments/assets/fe7027c5-a4ff-480a-8f5b-dc32fcb805f8" />

After checking the box:
<img width="902" alt="Screen Shot 2025-01-09 at 8 57 43 PM" src="https://github.com/user-attachments/assets/d106cc5a-6aae-4007-b61d-e159303c733d" />


Your PR basically does what the box is doing without it. Am I missing something?

**Comment by alanmels** [comment]
Hi @argiepiano, 

To understand the problem, please install the module on a fresh Backdrop and add an address field to a content type and choose NOT to make it required as shown below:

![Screenshot 2025-01-09 at 11 27 40 PM](https://github.com/user-attachments/assets/1f1ac2a4-ab73-466c-8c05-089360973ed1)

and go create a node of that content type, the fields are, nevertheless, required:

![Screenshot 2025-01-09 at 11 29 52 PM](https://github.com/user-attachments/assets/e3e1fafb-9f18-424b-8c73-7a4a6a9c07ac)

which contradicts the previous setting, that should be considered "global".

**Comment by alanmels** [comment]
I have only now noticed the additional "Make all fields optional (No validation - unsuitable for postal purposes)" option. Well, based on lots of reports for Drupal version and now here on Backdrop, the location of that field is not ideal at least. However, semantically it also contradicts with "Required field" option. Why should we have an additional option when it makes sense to choose if the field is required or not?!

**Comment by alanmels** [comment]
I've closed the PR as the "Make all fields optional" does work, however I still believe that field is just duplicating the function of the standard "Required field" box. It should've been either required or not. Why would anyone make it required and then check off "Make all fields optional"? Doesn't make sense.

**Comment by argiepiano** [comment]
Thanks @alanmels. Agreed that the placement of that option is not ideal. I had doubts about your PR also, since it will affect existing installations, suddenly allowing someone to submit just the country without any address. It's true that those sites should have made the field required in the first place, but a PR like this will create work for site admins (having to edit the widget to make the field required), many of whom may not be aware there is suddenly a problem they need to fix in their sites. There is also some gray area - you can always have the user select "None" in the country to make those subfields not required. There is a case for the widget requiring those subfields when you choose a country, even in a non-required address field, and making those subfields optional only when selecting "None" in the country. When you make the whole addressfield required, then you are not allowed to choose "None" in country, which makes sense.

I agree that there are lots of inconsistencies inherited from D7, but one has to be careful with changes to widget behaviors that will affect existing installations.

**Comment by argiepiano** [comment]
Will leave this issue open for future reference, though. Sorry for closing.

**Comment by alanmels** [comment]
The `Make all fields optional (No validation - unsuitable for postal purposes)` options works as expected, so let's clear the issue queue from long-standing discussions.
